### PR TITLE
Make TextFormFieldState.didChange change text fields value

### LINKED ISSUE #54472: TextFormFieldState.didChange should update text fields value
## Use case

There should be a way to update a text form field's value using only the form field state. The form field state actually exposes a method called didChange which is used to change the form field's internal value (which is used for validation etc.). Calling this method should not only update the form field's internal state but also the underlying text field.

## Proposal

Make TextFormFieldState.didChange change the underlying text field's value.

### PR DESCRIPTION
**Pull request by christian-muertz**
## Description

Calling a `TextFormField`s state didChange method currently only updates the internal state of the form field but not the actual value of the underlying text field. This PR makes the didChange method change the internal state as well as the actual text field value.

In addition, this PR removes a unnecessary setState call from the reset method.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/54472

## Tests

I added the following tests:

* A test to ensure that the didChange method updates the text fields value
* A test to ensure that the reset method updates the text fields value to the initialValue

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by christian-muertz** [comment]
@HansMuller

File: packages/flutter/lib/src/material/text_form_field.dart [hunk 1]
```diff
@@ -309,12 +309,18 @@ class _TextFormFieldState extends FormFieldState<String> {
...
+    if (_effectiveController.text != value)
+      _effectiveController.text = value;
+  }
+
   @override
   void reset() {
     super.reset();
-    setState(() {
-      _effectiveController.text = widget.initialValue;
-    });
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/text_form_field.dart:317]: super.reset does call setState(), so it's not strictly necessary. OTOH an extra setState() doesn't change anything and it saves the reader from needing to look at the superclass to discover that setState() is actually called..

File: packages/flutter/test/material/text_form_field_test.dart [hunk 2]
```diff
@@ -331,4 +331,48 @@ void main() {
     await tester.pump(const Duration(milliseconds: 300));
     expect(tapCount, 3);
   });
+
+  testWidgets('reset resets the text fields value to the initialValue', (WidgetTester tester) async {
+    await tester.pumpWidget(
```
Comment (HansMuller) [inline packages/flutter/test/material/text_form_field_test.dart:336]: Please include a comment here (and below)
```
// Regression test for https://github.com/flutter/flutter/issues/54472.
```

**Review by HansMuller — COMMENTED** [review]
This looks good assuming that it doesn't break any test-repo tests.

**Comment by HansMuller** [comment]
Still haven't been able to get a clean run of this PR through the internal test suite. So far the problems don't appear to have had anything to do with this PR. Still trying.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

I agree that the failures don't seem related, looked like an infrastructure thing.  I just re-ran them.

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**